modules/gapi/src/streaming/gstreamer/gstreamersource.cpp

```diff
@@ -184,10 +185,27 @@ void GStreamerSource::Priv::prepareVideoMeta()
             cv::util::throw_error(std::logic_error("Cannot query video width/height."));
         }
 
+        // Fill GstVideoInfo structure to work further with GstVideoFrame class.
+        if (!gst_video_info_from_caps(&m_videoInfo, prerollCaps)) {
+            cv::util::throw_error(std::logic_error("preroll sample has invalid caps."));
+        }
+        m_type = GST_VIDEO_INFO_FORMAT(&m_videoInfo);
         switch(m_outputType) {
             case GStreamerSource::OutputType::FRAME: {
                 // Construct metadata for media frame.
-                m_mediaFrameMeta = GFrameDesc { cv::MediaFormat::NV12, cv::Size(width, height) };
+                switch (m_type) {
+                    case GST_VIDEO_FORMAT_NV12: {
+                        m_mediaFrameMeta = GFrameDesc{ cv::MediaFormat::NV12, cv::Size(width, height) };
+                        break;
+                    }
+                    case GST_VIDEO_FORMAT_GRAY8: {
+                        m_mediaFrameMeta = GFrameDesc{ cv::MediaFormat::GRAY, cv::Size(width, height) };
+                        break;
+                    }
```

Comment:
Suggest to have corresponding asserts like  `GAPI_Assert(GST_VIDEO_INFO_N_PLANES(&m_videoInfo) == 2);` in each case (if `nPlanes == 1` for gray)

Reply:
done